src/BenchmarksWorkers/Workers/SignalRWorker.cs

```diff
@@ -0,0 +1,292 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Benchmarks.ClientJob;
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.AspNetCore.Sockets;
+using Newtonsoft.Json;
+
+namespace BenchmarksWorkers.Workers
+{
+    public class SignalRWorker : IWorker
+    {
+        public string JobLogText { get; set; }
+
+        private ClientJob _job;
+        private HttpClientHandler _httpClientHandler;
+        private List<HubConnection> _connections;
+        private List<IDisposable> _recvCallbacks;
+        private Timer _timer;
+        private List<int> _requestsPerConnection;
+        private List<List<double>> _latencyPerConnection;
+        private Stopwatch _workTimer = new Stopwatch();
+        private bool _stopped;
+
+        public SignalRWorker(ClientJob job)
+        {
+            _job = job;
+
+            Debug.Assert(_job.Connections > 0, "There must be more than 0 connections");
+
+            // Configuring the http client to trust the self-signed certificate
+            _httpClientHandler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+            };
+
+            var jobLogText =
+                $"[ID:{_job.Id} Connections:{_job.Connections} Duration:{_job.Duration} Method:{_job.Method} ServerUrl:{_job.ServerBenchmarkUri}";
+
+            if (_job.Headers != null)
+            {
+                jobLogText += $" Headers:{JsonConvert.SerializeObject(_job.Headers)}";
+            }
+
+            TransportType transportType = default;
+            if (_job.ClientProperties.TryGetValue("TransportType", out var transport))
+            {
+                transportType = Enum.Parse<TransportType>(transport);
+                jobLogText += $" TransportType:{transportType}";
+            }
+
+            jobLogText += "]";
+            JobLogText = jobLogText;
+
+            CreateConnections(transportType);
+        }
+
+        public async Task StartAsync()
+        {
+            // start connections
+            var tasks = new List<Task>(_connections.Count);
+            foreach (var connection in _connections)
+            {
+                tasks.Add(connection.StartAsync());
+            }
+
+            await Task.WhenAll(tasks);
+
+            _job.State = ClientState.Running;
+            _job.LastDriverCommunicationUtc = DateTime.UtcNow;
+
+            // SendAsync will return as soon as the request has been sent (non-blocking)
+            await _connections[0].SendAsync("Echo", _job.Duration + 1);
+            _workTimer.Start();
+            _timer = new Timer(StopClients, null, TimeSpan.FromSeconds(_job.Duration), Timeout.InfiniteTimeSpan);
```

Comment:
Wait, we have a timer that runs once?

Reply:
A new worker is created for every job, and the timer is to stop the worker when the job is done